File: dawn_python/fusion_tree.py
class FusionTree:
    def __init__(self, value=0, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right

    @classmethod
    def empty(cls):
        return cls(0)

    def is_empty(self):
        return self.value == 0 and self.left is None and self.right is None

    def insert(self, x):
        if self.is_empty():
            return FusionTree(x)
        else:
            left = self.left.insert(x) if self.left else None
            right = self.right
            return FusionTree(self.value, left, right)

    def find(self, x):
        if self.is_empty():
            return 0
        elif self.value == x:
            left_sum = self.left.find(x) if self.left else 0
            right_sum = self.right.find(x) if self.right else 0
            return self.value + left_sum + right_sum
        else:
            left_sum = self.left.find(x) if self.left else 0
            right_sum = self.right.find(x) if self.right else 0
            return left_sum + right_sum

    @staticmethod
    def union(left, right):
        if left.is_empty():
            return right
        elif right.is_empty():
            return left
        else:
            new_left = FusionTree.union(left.left, right.left)
            new_right = FusionTree.union(left.right, right.right)
            return FusionTree(left.value + right.value, new_left, new_right)

File: dawn_cpp/fusion_tree.cpp
#include <memory>

template <typename T>
struct FusionTree {
    T value;
    std::unique_ptr<FusionTree<T>> left, right;

    FusionTree(T val) : value(val), left(nullptr), right(nullptr) {}

    static FusionTree<T> empty() {
        return FusionTree<T>(T());
    }

    bool is_empty() const {
        return value == T() && !left && !right;
    }

    FusionTree<T> insert(T x) {
        if (is_empty()) {
            return FusionTree<T>(x);
        }
        auto new_left = left->insert(x);
        return FusionTree<T>(value, std::make_unique<FusionTree<T>>(std::move(new_left)), std::move(right));
    }

    T find(T x) const {
        if (is_empty()) {
            return T();
        }
        if (value == x) {
            T left_sum = left ? left->find(x) : T();
            T right_sum = right ? right->find(x) : T();
            return value + left_sum + right_sum;
        }
        T left_sum = left ? left->find(x) : T();
        T right_sum = right ? right->find(x) : T();
        return left_sum + right_sum;
    }

    static FusionTree<T> union_(const FusionTree<T>& left, const FusionTree<T>& right) {
        if (left.is_empty()) {
            return right;
        }
        if (right.is_empty()) {
            return left;
        }
        auto new_left = union_(*left.left, *right.left);
        auto new_right = union_(*left.right, *right.right);
        return FusionTree<T>(left.value + right.value, std::make_unique<FusionTree<T>>(std::move(new_left)), std::make_unique<FusionTree<T>>(std::move(new_right)));
    }
};